Add possibility to include the end of the temporal FilterRange

rduivenvoorde opened the pull request:
## Description

Default temporal filtering (of vector features) in most software will
INclude the begin/start of the filtering range, but EXclude the end of it.
So: if you filter with say a filter of 1900-1910, by default features with
1900 will be selected, but with 1910 will NOT (exclude end).

Some context:
- https://github.com/qgis/QGIS/pull/40989 (on the end comments)
- leading to: https://github.com/qgis/QGIS/issues/44589

This commit adds the possibility (in the Temporal Features tab of the
vector layer properties) to choose the 'limitMode'.
The limitmode can now also be set to INclude begin AND INclude end.
So in the above example, also the features with 1910 will be selected.

The limitMode is written to the xml project file as temporal attribute.

Below the added ComboBox

![Screenshot-20210812171016-974x216](https://user-images.githubusercontent.com/731673/129221545-16f0abd7-164b-403e-9369-0ea781c5c05b.png)

Feel free to comment on code/naming or place of combobox :-)

A small test project with a csv loaded as temporal layers:

[temporalLimitTest.zip](https://github.com/qgis/QGIS/files/6976618/temporalLimitTest.zip)

rduivenvoorde committed:
Adding the possibility to include the end of the temporal filtering TimeRange

Default temporal filtering (of vector features) in most software will
INclude the begin/start of the filtering range, but EXclude the end of it.
So: if you filter with say a filter of 1900-1910, by default features with
1900 will be selected, but with 1910 will NOT (exclude end).

This commit adds the possibility (in the Temporal Features tab of the
vector layer properties) to choose the 'limitMode'.
The limitmode can now also be set to INclude begin AND INclude end.
So in the above example, also the features with 1910 will be selected.

The limitMode is written to the xml project file as temporal attribute.

rduivenvoorde committed:
Sipify and indentation

agiudiceandrea commented:
I think the new option "Include Start, Include End" should affect the inclusion or exclusion of the "end_field" of the "feature range", not the QgsDateTimeRange map temporal range.

So it should affect the second part of e.g. the filter string `"(%1 %2 %3 OR %1 IS NULL) AND (%4 > %5 OR %4 IS NULL)"` and not the first part.

For example, this
```c++
    case ModeFeatureDateTimeStartAndEndFromFields:
    {
      if ( !mStartFieldName.isEmpty() && !mEndFieldName.isEmpty() )
        return QStringLiteral( "(%1 %2 %3 OR %1 IS NULL) AND (%4 > %5 OR %4 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
               filterRange.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
               dateTimeExpressionLiteral( filterRange.end() ),
               QgsExpression::quotedColumnRef( mEndFieldName ),
               dateTimeExpressionLiteral( filterRange.begin() ) );
      }
```
would became something like:
```c++
    case ModeFeatureDateTimeStartAndEndFromFields:
    {
      if ( !mStartFieldName.isEmpty() && !mEndFieldName.isEmpty() )
        return QStringLiteral( "(%1 %2 %3 OR %1 IS NULL) AND (%4 %5 %6 OR %4 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
               filterRange.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
               dateTimeExpressionLiteral( filterRange.end() ),
               QgsExpression::quotedColumnRef( mEndFieldName ),
               limitMode() == QgsVectorLayerTemporalProperties::ModeIncludeBeginIncludeEnd ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
               dateTimeExpressionLiteral( filterRange.begin() ) );
      }
```
or
```c++
    case ModeFeatureDateTimeStartAndEndFromFields:
    {
      if ( !mStartFieldName.isEmpty() && !mEndFieldName.isEmpty() )
        return QStringLiteral( "(%1 %2 %3 OR %1 IS NULL) AND (%4 %5 %6 OR %4 IS NULL)" ).arg( QgsExpression::quotedColumnRef( mStartFieldName ),
               filterRange.includeEnd() ? QStringLiteral( "<=" ) : QStringLiteral( "<" ),
               dateTimeExpressionLiteral( filterRange.end() ),
               QgsExpression::quotedColumnRef( mEndFieldName ),
               ( limitMode() == QgsVectorLayerTemporalProperties::ModeIncludeBeginIncludeEnd && filterRange.includeBeginning() ) ? QStringLiteral( ">=" ) : QStringLiteral( ">" ),
               dateTimeExpressionLiteral( filterRange.begin() ) );
      }
```

nyalldawson commented:
+1 to @agiudiceandrea's concern

And thanks @rduivenvoorde for taking this on!

rduivenvoorde commented:
both Thanks: yes I will have a closer look, and fix the tests...

agiudiceandrea commented on python/core/auto_generated/vector/qgsvectorlayertemporalproperties.sip.in:
```diff
@@ -74,6 +74,12 @@ The ``enabled`` argument specifies whether the temporal properties are initially
       ModeRedrawLayerOnly,
     };
 
+    enum LimitMode
+    {
+        ModeIncludeBeginExcludeEnd,
+        ModeIncludeBeginIncludeEnd,
+    };
```
```suggestion
    enum LimitMode
    {
      ModeIncludeBeginExcludeEnd,
      ModeIncludeBeginIncludeEnd,
    };
```

rduivenvoorde commented:
@agiudiceandrea I tried to change the logic so the 'limit inclusion' moved to the second part, but in practice that does make the actual filter outcome not more intuitive...

Let me try to explain:
- I have the dataset attached above: features with one hour events: 1-2, 2-3, 3-4 etc
- stepping in that set in one hour timeframes

Original setup (pre PR) the 1-2 filter (in which t_start is the start field and t_end is the end field) became:
` ("t_start" < make_datetime(2020,8,26,2,0,0) OR "t_start" IS NULL) AND ("t_end" > make_datetime(2020,8,26,1,0,0) OR "t_end" IS NULL)`
Resulting in:
- step 1 = 1-2 will only show you 1-2

Current PR with the 'end limit including' the 1-2 filter becomes:
`("t_start" <= make_datetime(2020,8,26,2,0,0) OR "t_start" IS NULL) AND ("t_end" > make_datetime(2020,8,26,1,0,0) OR "t_end" IS NULL)`
Resulting in: 
- step 1 = 1-2 will show you 1-2 AND 2-3
- step 2 = 2-3 will show you 2-3 AND 3-4

Taking your proposal, the 1-2 filter becomes:
`("t_start" < make_datetime(2020,8,26,2,0,0) OR "t_start" IS NULL) AND ("t_end" >= make_datetime(2020,8,26,1,0,0) OR "t_end" IS NULL)`
Resulting in:
- step 1 = 1-2 will show you only 1-2
- step 2 = 2-3 will show you 1-2 AND 2-3

It's all in 'what does the user mean when he says: INCLUDING END':
- is it: the end of the FILTER FRAME (so the frame 1-2 will 'touch' (and thus select) 2-3
- or is it the end of the FEATURE EVENT END (or duration END) (so the feature 1-2 will be 'touched' both by 0-1 AND 1-2 AND 2-3 (given we already did have a closed start and open end)...

I actually think/thought that we were not talking about including the END of the feature but about using the END of the filter frame to check 'touching' events.

Side note: not having 'features with events with a duration', but 'events with a momentarily event (only one timestamp)' you can not talk about 'including the end limit of the feature' while you can still talk about 'including the end limit of the filtering time frame'

Correct me if I am plain wrong (I'm trying to make things clear, but happen to make things more blurry, ask @nyalldawson), but my impression is that we are talking about the inclusion of the end of the filtering frame (the start of the filtering we already always include, else you would get missing features).

What do you think?

rduivenvoorde commented:
Ok, I'm on both your side now... after some experimenting/thinking :-)

rduivenvoorde committed:
Sipify all

rduivenvoorde committed:
Add debug line in renderer to see the temporal filter strings easier

rduivenvoorde committed:
Adding the possibility to include the end of the feature range

 Default temporal filtering (of vector features) in most software will
 INclude the begin/start of the feature range, but EXclude the end of it.
 So: if you filter with say a filter of 1900-1910, by default features with
 1900 will be selected, but with 1910 will NOT (exclude end).

 This commit adds the possibility (in the Temporal Features tab of the
 vector layer properties) to choose the 'limitMode'.
 The limitmode can now also be set to INclude begin AND INclude end.
 So in the above example, also the features with 1910 will be selected.

 The limitMode is written to the xml project file as temporal attribute.

rduivenvoorde commented:
Adding data/project here to to do some testing (see screenshot below):

[TempLimitProject.zip](https://github.com/qgis/QGIS/files/7038353/TempLimitProject.zip)

Also note that I added a debug line in the rendering code, to more easily check the actual query created (when you Filterh the Application Output with 'Rending with Temporal Filter' you will see all temporal queries like below:

![Screenshot-20210824120039-1497x1280](https://user-images.githubusercontent.com/731673/130598350-1dddd0f4-0864-43a7-819c-f3f2c0f03a01.png)

I probably also have to add some tests?

rduivenvoorde committed:
Oops, forgot this one

rduivenvoorde committed:
Add tests for the inclusion of the Feature event end

CoryAlbrecht commented:
I have been using this branch for the past few days with Include End turned on, and it has been working for me.

rduivenvoorde commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -273,6 +275,11 @@ def testSingleFieldMode(self):
         self.assertEqual(props.createFilterString(context, range), '("start_field" > make_datetime(2019,3,1,11,12,13) AND "start_field" < make_datetime(2020,5,6,8,9,10)) OR "start_field" IS NULL')
 
+        # since 3.22 there is also the option to include the end of the feature event
+        props.setLimitMode(QgsVectorLayerTemporalProperties.LimitMode.ModeIncludeBeginIncludeEnd)
```
Yes, sure. I'll do that

rduivenvoorde commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -99,6 +99,7 @@ QgsVectorLayerRenderer::QgsVectorLayerRenderer( QgsVectorLayer *layer, QgsRender
     QgsVectorLayerTemporalContext temporalContext;
     temporalContext.setLayer( layer );
     mTemporalFilter = qobject_cast< const QgsVectorLayerTemporalProperties * >( layer->temporalProperties() )->createFilterString( temporalContext, context.temporalRange() );
+    qDebug() << "Rendering with Temporal Filter: " << mTemporalFilter;
```
If I make that a QgsDebugMsg level .. high. Is it ok to leave it then?

rduivenvoorde committed:
Remove qDebug call

rduivenvoorde committed:
Add QgsDebugMsg to QgsVectorLayerRenderer

This way it is easier to see the filter query used for all different
settings and options

rduivenvoorde committed:
Add diagrams to tests

rduivenvoorde commented:
IF 4a88360 is not wanted, feel free to remove it

rduivenvoorde commented:
Added some diagrams to the test. Current status is ok to merge for me.

nyalldawson commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -99,6 +99,7 @@ QgsVectorLayerRenderer::QgsVectorLayerRenderer( QgsVectorLayer *layer, QgsRender
     QgsVectorLayerTemporalContext temporalContext;
     temporalContext.setLayer( layer );
     mTemporalFilter = qobject_cast< const QgsVectorLayerTemporalProperties * >( layer->temporalProperties() )->createFilterString( temporalContext, context.temporalRange() );
+    QgsDebugMsg( "Rendering with Temporal Filter: " + mTemporalFilter );
```
```suggestion
    QgsDebugMsgLevel( "Rendering with Temporal Filter: " + mTemporalFilter, 2 );
```

nyalldawson replied:
(level 1 is reserved for "error" messages)

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -273,6 +275,22 @@ def testSingleFieldMode(self):
+        #                    [---.---]                     .            (true)
+        #                [-------]                         .            (true)
+        #          [-------]     .                         .            (false)
+        # => start of feature < end of range AND start of feature + duration > start of range
```
Looking at the diagram, isn't this 

```suggestion
        # => start of feature < end of range AND start of feature + duration >= start of range
```
?

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -273,6 +275,22 @@ def testSingleFieldMode(self):
+        #                [-------]                         .            (true)
+        #          [-------]     .                         .            (false)
+        # => start of feature < end of range AND start of feature + duration > start of range
+        # OR start of feature <= end of range AND start of feature > start of range - duration
```
```suggestion
        # OR start of feature < end of range AND start of feature >= start of range - duration
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -451,6 +469,7 @@ def testDualFieldMode(self):
         range = QgsDateTimeRange(QDateTime(QDate(2019, 3, 4), QTime(11, 12, 13)), QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10)), includeBeginning=False)
         self.assertEqual(props.createFilterString(context, range), '("start_field" <= make_datetime(2020,5,6,8,9,10) OR "start_field" IS NULL) AND ("end_field" > make_datetime(2019,3,4,11,12,13) OR "end_field" IS NULL)')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -464,6 +483,21 @@ def testDualFieldMode(self):
+        #                [-------]                         .            (true)
+        #          [-------]     .                         .            (false)
+        #
+        # => start of feature < end of range AND end of feature >= start of range
```
agreed

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -501,6 +535,7 @@ def testDualFieldMode(self):
         # => start of feature <= end of range
         self.assertEqual(props.createFilterString(context, range), '"start_field" <= make_datetime(2020,5,6,8,9,10) OR "start_field" IS NULL')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -258,6 +258,8 @@ def testSingleFieldMode(self):
         # OR start of feature <= end of range AND start of feature > start of range - duration
         self.assertEqual(props.createFilterString(context, range), '("start_field" > make_datetime(2019,3,1,11,12,13) AND "start_field" <= make_datetime(2020,5,6,8,9,10)) OR "start_field" IS NULL')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -549,8 +584,9 @@ def testDualFieldMode(self):
         # => end of feature > start of range
         self.assertEqual(props.createFilterString(context, range), '"end_field" > make_datetime(2019,3,4,11,12,13) OR "end_field" IS NULL')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -559,6 +595,19 @@ def testDualFieldMode(self):
+        #                --------]                         .            (true)
+        #                -----]                            .            (false)
+        #
+        # => end of feature >= start of range
```
agreed

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -627,6 +676,7 @@ def testStartAndDurationMode(self):
         self.assertEqual(props.createFilterString(context, range),
                          '("start_field" <= make_datetime(2020,5,6,8,9,10) OR "start_field" IS NULL) AND (("start_field" + make_interval(0,0,0,0,0,0,"duration"/1000) > make_datetime(2019,3,4,11,12,13)) OR "duration" IS NULL)')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -642,6 +692,21 @@ def testStartAndDurationMode(self):
                          '("start_field" < make_datetime(2020,5,6,8,9,10) OR "start_field" IS NULL) AND (("start_field" + make_interval(0,0,0,0,0,0,"duration"/1000) > make_datetime(2019,3,4,11,12,13)) OR "duration" IS NULL)')
+        # map range              [-------------------------)
+        # feature ranges         .                         . [-------]  (false)
+        #                        .                         [-------]    (true)
```
There's no overlap here:
```suggestion
        #                        .                         [-------]    (false)
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -642,6 +692,21 @@ def testStartAndDurationMode(self):
+        #                [-------]                         .            (true)
+        #          [-------]     .                         .            (false)
+        #
+        # => start of feature <= end of range AND start + duration > start of range
```
From the updated diagram, this should be:
```suggestion
        # => start of feature < end of range AND start + duration >= start of range
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -741,6 +806,7 @@ def testExpressionMode(self):
         # => start expression <= end of range AND end expression > start of range
         self.assertEqual(props.createFilterString(context, range), '((to_datetime("my string field",  \'yyyy MM dd hh::mm:ss\')") <= make_datetime(2020,5,6,8,9,10)) AND ((to_datetime("my end field",  \'yyyy MM dd hh::mm:ss\')") > make_datetime(2019,3,4,11,12,13))')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -754,6 +820,20 @@ def testExpressionMode(self):
+        #                [-------]                         .            (true)
+        #          [-------]     .                         .            (false)
+        #
+        # => start expression <= end of range AND end expression > start of range
```
From the diagram this should be:

```suggestion
        # => start expression < end of range AND end expression >= start of range
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -839,6 +919,7 @@ def testExpressionMode(self):
         # => end of feature > start of range
         self.assertEqual(props.createFilterString(context, range), '(to_datetime("my end field",  \'yyyy MM dd hh::mm:ss\')") > make_datetime(2019,3,4,11,12,13)')
 
+        # THIS is the QGIS default: using a range with includeBeginning=true en includeEnd=false
```
```suggestion
        # THIS is the QGIS default: using a range with includeBeginning=true and includeEnd=false
```

nyalldawson commented on tests/src/python/test_qgsvectorlayertemporalproperties.py:
```diff
@@ -849,6 +930,16 @@ def testExpressionMode(self):
+        #                --------.--------------------]    .            (true)
+        #                --------]                         .            (true)
+        #                -----]                            .            (false)
+        #        # => end of feature >= start of range
```
agreed

nyalldawson commented:
@rduivenvoorde from the diagrams it looks like there's some logic errors here. Can you review my comments and see if my suggestions are correct? (If so, the generated expressions will need to be updated accordingly)

rduivenvoorde commented:
@nyalldawson argh, sorry I'll have a more thorough look. I took the copy/paste road too easy.

rduivenvoorde committed:
Update src/core/vector/qgsvectorlayerrenderer.cpp

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

rduivenvoorde committed:
Fix includeEnd logic for other then Start/End from fields

rduivenvoorde committed:
Fix tests according to fixed logic

rduivenvoorde commented:
@nyalldawson very VERY sharp!!! Compliments for seeing my flaws! You were definitely right:
- the logic for the start/end from fields were OK
- but the expressions + with duration ones were NOT ok
I fixed the logic AND compared the output using the test project above, which (I think/hope) are now good.
Also fixed the tests.
Thanks

nyalldawson committed:
Update tests/src/python/test_qgsvectorlayertemporalproperties.py

nyalldawson commented:
Looks great, thanks for your hard work here @rduivenvoorde !

CoryAlbrecht commented:
@rduivenvoorde I've been noticing some weird things in the last couple updates to this branch.

One is that when I edit an attribute that is not the start or end date fields, the feature disappears from the canvas view. The only thing that makes it reappear is saving the layer.

Also, when I change the layer styles, save the project back to the PostGIS database, close qgis, restart, and reload the project, the style changes were not saved along with the rest of the project. Even choosing Style→Save as default on the Symbology properties tab doesn't make the new changes permanent.

They don't seem to me to be things related to your patches?

rduivenvoorde commented:
@CoryAlbrecht mmm, not sure... what the temporal controller does is create a 'filter string' to go over the features and only paint the features that return 'true' from this filter.... the result of my patches are tiny changes (> to >= etc)...
BUT if for example the painting *during the editing* does use/have another query at that moment and thus a different result, that I could imagine that your features 'disappear'... 
So 'save' will make them appear again?
@nyalldawson does this ring a bell with you? Is it possible that during editing there is another code path (for example one that does NOT use 'setLimitMode' yet)?
Also: is it possible I have to 'save' the limitmode setting into the project file?

rduivenvoorde commented:
@CoryAlbrecht I tried to create a tiny test by putting some points in a geopackage, but I cannot confirm.
Can you try to see if you can replicate with test set below?

[test.zip](https://github.com/qgis/QGIS/files/7139287/test.zip)

or else upload a tiny dataset here with which I can replicate.
You are sure it is not style related?

nyalldawson commented:
@rduivenvoorde 
> does this ring a bell with you? Is it possible that during editing there is another code path (for example one that does NOT use 'setLimitMode' yet)?

No, the exact same filters will work fine with unsaved features. (just how rule based renderer filters work fine regardless of whether the features are saved/unsaved).

There's something else happening here unrelated to this which we'd need more steps to reproduce.
